pyteal/ast/opup.py

```diff
@@ -0,0 +1,163 @@
+        elif mode == OpUpMode.OnCall:
+            self.target_app_id_slot = ScratchSlot()
+        else:
+            raise TealInputError("Invalid OpUp mode provided")
+
+        self.mode = mode
+
+        # A budget buffer is necessary to deal with an edge case of ensure_budget():
+        #   if the current budget is equal to or only slightly higher than the
+        #   required budget then it's possible for ensure_budget() to return with a
+        #   current budget less than the required budget. The buffer prevents this
+        #   from being the case.
+        self.buffer = Int(50)
+
+    def _create_app(self) -> Expr:
+        self.target_app_id = self.target_app_id_slot.load()
+        return Seq(
+            InnerTxnBuilder.Begin(),
+            InnerTxnBuilder.SetFields(
+                {
+                    TxnField.type_enum: TxnType.ApplicationCall,
+                    TxnField.approval_program: ON_CALL_APP,
+                    TxnField.clear_state_program: ON_CALL_APP,
+                }
+            ),
+            InnerTxnBuilder.Submit(),
+            self.target_app_id_slot.store(Gitxn[0].created_application_id()),
+        )
+
+    def _delete_app(self, app_id: Expr) -> Expr:
+        return Seq(
+            InnerTxnBuilder.Begin(),
+            InnerTxnBuilder.SetFields(
+                {
+                    TxnField.type_enum: TxnType.ApplicationCall,
+                    TxnField.on_completion: OnComplete.DeleteApplication,
+                    TxnField.application_id: app_id,
+                }
+            ),
+            InnerTxnBuilder.Submit(),
+        )
+
+    def _ensure_budget_expr(self, required_budget: Expr) -> Expr:
+        return While(Global.opcode_budget() < required_budget).Do(
+            Seq(
+                InnerTxnBuilder.Begin(),
+                InnerTxnBuilder.SetFields(
+                    {
+                        TxnField.type_enum: TxnType.ApplicationCall,
+                        TxnField.application_id: self.target_app_id,
+                    }
+                ),
+                InnerTxnBuilder.Submit(),
+            )
+        )
+
+    def ensure_budget(self, required_budget: Expr) -> Expr:
+        """Ensure that the budget will be at least the required_budget."""
+        require_type(required_budget, TealType.uint64)
+
+        buffered_budget = required_budget + self.buffer
+        if self.mode == OpUpMode.OnCall:
+            return If(Global.opcode_budget() < buffered_budget).Then(
+                Seq(
+                    self._create_app(),
+                    self._ensure_budget_expr(buffered_budget),
+                    self._delete_app(self.target_app_id),
+                )
+            )
+
+        return self._ensure_budget_expr(buffered_budget)
+
+    def _maximize_budget_expr(self, fee: Expr) -> Expr:
+        i = ScratchVar(TealType.uint64)
+        n = fee / Int(1000)
+        return For(i.store(Int(0)), i.load() < n, i.store(i.load() + Int(1))).Do(
```

Comment:
@algoidurovic Is it accurate that `maximize_budget` presumes a 1 micro Algo transaction fee?

* I have limited experience here though I know from https://developer.algorand.org/docs/get-details/transactions/#fees that transaction fees are variable.
* I did a brief scan and I don't see a way to gauge the transaction fee without attempting the transaction.
* If I'm following along, then I'm wondering if there's a specific use case that prompted the _maximize_budget_ approach or if it's been added for convenience?
  * I see the intended convenience.  _If_ it's intended as a convenience and _if_ there's there's no reasonable way to account for the dynamic transaction fees, I'm thinking the functionality can be added in a later iteration.

Reply:
Summarizing a verbal with @algoidurovic:
* There's 2 levels of configurable fees:
  * Congestion fees
  * Min transaction fees configured via consensus parameter
* Igor to investigate:
  * Replace `Int(1000)` with a min transaction fee field to catch consensus parameter changes.
  * Identify if congestion fees apply to inner transactions.  If yes, we ought to highlight that the method _can_ fail at evaluation time by either (or both) changing method name and clarifying docs.

Reply:
Thanks to @algoidurovic for pointing out that inner transactions are _not_ subject to congestion fees because they're _not_ transmitted over the network.

https://github.com/algorand/pyteal/pull/274/commits/7152af649c6ab0d5a1a6690fcaee95268240b3ea brings in the feedback to factor out the min txn fee constant.    

Resolving.